pkg/network/status/status.go

```diff
@@ -395,3 +394,37 @@ func (m *Prober) updateStates(ingressState *ingressState, podState *podState) {
 		}
 	}
 }
+
+func (m *Prober) probeVerifier(item *workItem) prober.Verifier {
+	return func(r *http.Response, _ []byte) (bool, error) {
+		// In the happy path, the probe request is forwarded to Activator or Queue-Proxy and the response (HTTP 200)
+		// contains the "K-Network-Hash" header that can be compared with the expected hash. If the hashes match,
+		// probing is successful, if they don't match, a new probe will be sent later.
+		// An HTTP 404 is expected in the case of the creation of a new Knative service because the rules will
+		// not be present in the Envoy config until the new VirtualService is applied.
+		// No information can be extracted from any other scenario (e.g. HTTP 302), therefore in that case,
+		// probing is assumed to be successful because it is better to say that an Ingress is Ready before it
+		// actually is Ready than never marking it as Ready. It is best effort.
+		switch r.StatusCode {
+		case http.StatusOK:
+			hash := r.Header.Get(network.HashHeaderName)
+			switch hash {
+			case "":
+				m.logger.Errorf("Probing of %s abandoned, IP: %s:%s: the response doesn't contain the %q header",
+					item.url, item.podIP, item.podPort, network.HashHeaderName)
+				return true, nil
+			case item.ingressState.hash:
+				return true, nil
+			default:
+				return false, fmt.Errorf("unexpected hash: want %q, got %q", item.ingressState.hash, hash)
+			}
+		case http.StatusNotFound:
```

Comment:
Can we also include 503s here, since those were also observed in early requests?

Reply:
503 should not happen when the VirtualService has not been applied in the config.
503 are when Envoy is overloaded.

Reply:
I have seen "503" with response flag "NR" (No Route) from Envoy which seems to be essentially a 404.

Reply:
Fair enough, plus HTTP 503 is supposed to be for **temporary** issues, so it should be re-tried.